Sources/NIOCertificateHotReloading/NIOCertificateHotReloading.swift

```diff
@@ -0,0 +1,319 @@
+    }
+
+    private let refreshInterval: TimeAmount
+    private let certificateDescription: CertificateDescription
+    private let privateKeyDescription: PrivateKeyDescription
+    private let state: NIOLockedValueBox<CertificateKeyPair?>
+
+    /// A `NIOSSLContextConfigurationOverride` that will be used as part of the NIO application's TLS configuration.
+    /// Its certificate and private key will be kept up-to-date via the reload mechanism the ``TimedCertificateReloader``
+    /// implementation provides.
+    public var sslContextConfigurationOverride: NIOSSLContextConfigurationOverride {
+        get {
+            var override = NIOSSLContextConfigurationOverride()
+            guard let certificateKeyPair = self.state.withLockedValue({ $0 }) else {
+                return override
+            }
+            override.certificateChain = [certificateKeyPair.certificate]
+            override.privateKey = certificateKeyPair.privateKey
+            return override
+        }
+    }
+
+    /// Initialize a new ``TimedCertificateReloader``.
+    /// - Parameters:
+    ///   - refreshInterval: The interval at which attempts to update the certificate and private key should be made.
+    ///   - certificateDescription: A ``CertificateDescription``.
+    ///   - privateKeyDescription: A ``PrivateKeyDescription``.
+    public init(
+        refreshInterval: TimeAmount,
+        certificateDescription: CertificateDescription,
+        privateKeyDescription: PrivateKeyDescription
+    ) {
+        self.refreshInterval = refreshInterval
+        self.certificateDescription = certificateDescription
+        self.privateKeyDescription = privateKeyDescription
+        self.state = NIOLockedValueBox(nil)
+
+        // Immediately try to load the configured cert and key to avoid having to wait for the first
+        // reload loop to run.
+        self.reloadPair()
+    }
+
+    /// Initialize a new ``TimedCertificateReloader``.
+    /// - Parameters:
+    ///   - refreshInterval: The interval at which attempts to update the certificate and private key should be made.
+    ///   - certificateDescription: A ``CertificateDescription``.
+    ///   - privateKeyDescription: A ``PrivateKeyDescription``.
+    @available(macOS 13, iOS 16, tvOS 16, watchOS 9, *)
+    public init(
+        refreshInterval: Duration,
+        certificateDescription: CertificateDescription,
+        privateKeyDescription: PrivateKeyDescription
+    ) {
+        self.init(
+            refreshInterval: TimeAmount(refreshInterval),
+            certificateDescription: certificateDescription,
+            privateKeyDescription: privateKeyDescription
+        )
+    }
+
+    /// A long-running method to run the ``TimedCertificateReloader`` and start observing updates for the certificate and
+    /// private key pair.
+    /// - Important: You *must* call this method to get certificate and key updates.
+    public func run() async throws {
+        while !Task.isCancelled {
```

Comment:
Might be better to do `while true` to keep behavior consistent regardless of when exactly cancelation happens. Otherwise sometimes you might end up with a clean return and with a thrown cancelation error otherwise.

Reply:
Good point. Makes me wonder whether this should throw at all because with that change it can never return cleanly. (Not sure this actually matters much.)

Reply:
Yeah it'd be good to get more guidance on the expected behavior - should services return or throw CancellationError. I've seen us do both.

Reply:
Yeah good callout - I'll switch to `while true`.

With regards to whether it should be throwing or we should return, a (maybe useful?) data point, `ServiceLifecycle` will throw `CancellationError` if you try gracefully shutting down a service that's already been cancelled - even if the service itself didn't throw `CancellationError` when it was cancelled. Also, many of the async APIs Apple provides (like `Task.sleep` in this case, or async iterators' `next`) will handle cancellation by throwing a `CancellationError` too. 

I think throwing out of this `Task.sleep` is particularly useful because you don't want to wait for the whole timeout period, which could be minutes, to finish before the reloading task is cancelled. I do think it's the right thing to do here.